net-im/slack++/slack++-0.recipe

```diff
@@ -0,0 +1,56 @@
+SUMMARY="Slack Enterprise chat and services on the command line"
+DESCRIPTION="slack++ is an ncurses based client for the Searchable Log of All Conversation and Knowledge"
+HOMEPAGE="https://github.com/ga2arch/slack"
+COPYRIGHT="2018 Federico Di Pierro"
+LICENSE="GNU GPL v3"
+REVISION="1"
+srcGitRev="f085f2a269d012c876e128db74d350754ec0ca30"
+SOURCE_URI="https://github.com/ga2arch/slack/archive/$srcGitRev.tar.gz"
+CHECKSUM_SHA256="2b42d26328e666dbbd431e8f389234eb19fd56f4ad65997d3d442b0cb96a4b6a"
+SOURCE_DIR="slack-$srcGitRev"
+
+ARCHITECTURES="!x86_gcc2 x86 x86_64"
+SECONDARY_ARCHITECTURES="x86"
+PROVIDES="
```

Comment:
I'm not sure how this change to secondary architectures is useful? What I am missing there?

Reply:
I do agree that it won't make any diference to most users.
But to those users who have built x86_64 hybrid with support for x86 secondary arch, having `slack++_x86` as a buildable recipe is of no use. They already have `slack++` for x86_64 primary arch.
`slack++_x86` is only needed on x86_gcc2 hybrid.

BTW, you can improve the recipe a little bit by [using the commandBinDir User Variable](https://github.com/haikuports/haikuports/wiki/HaikuPorter-Guidelines#using-the-commandbindir-user-variable) and making this change in `PROVIDES`:
```diff
- 	cmd:slack++$secondaryArchSuffix = $portVersion
+	cmd:slack++ = $portVersion
```

Reply:
I don't buy it. We already had similar workarounds, to mark x86_gcc2 as not-broken when build for secondary architectures. Haikuporter was changed to remove this. Why introduce something like this again? What if x86_64 gets support for x86? Are you going to remove it from all recipes?

Reply:
Of course not!
We only need to do this for recipes which do not provide any `lib:lib*` in `PROVIDES`.

Reply:
@cb88, have you had a look in `/boot/home/haikuports/repository/`?
This directory contains a bunch of files. If a recipe, e.g. `slack++`, is available for e.g. x86_64 primary arch and x86 secindary arch, and if you happen to be running x86_64 hybrid and have `x86` in your `haikuports.conf` then you will have one file for `slack++` and another for `slack++_x86`, The first one is welcome. The second one is totally unneeded.
Why would we give up making a change that only adds 2 lines but saves us an unneeded file in the `repository` directory?

I'm not making this change on existing recipes if I don't have another real reason to edit them.
But on **new** recipes which do not provide any library, not adding the condition for the secondary arch is IMO a bad practice.

Reply:
Shall we vote?
Comments welcome.

Reply:
A recipe should just encode the minimum information that it needs to here.

Just consider how ugly this could get if more Haiku architectures are added.... and there are binary translators on those arches that allow running other arches packages etc... on ARM for instance people may want to use qemu-user-static-i386 to run old BeOS apps and they could just install the packages as anohter secondary arch.

Reply:
> A recipe should just encode the minimum information that it needs to here.

The changes I'm requesting will:
1) allow x86_gcc2h users to type `slack++` instead of `slack++-x86`.
2) contribute to keep the `repository` directory free of unneeded files.

If you want point 1 then you need the test, otherwise the recipe will not be 100% valid because, on x86_84 hybrid, `cmd:slack++` will be provided by both `slack++` and `slack++_x86`. To avoid this conflict there are two solutions: Adding a conditional `CONFLICTS`, or adding the condition for the secondary arch.

> Just consider how ugly this could get if more Haiku architectures are added.... 

It is very unlikely that a new arch would be supported as a secondary arch on any of the primary arches were `slack++` builds, i.e. x86 and x86_64, so the test won't need to be changed.

Now, once the ARM port is ready, if `slack++` builds on it, then there will be no need to support x86 as a secondary arch for ARM, hence no need to change the test, which will still be:
```bash
if [ "$targetArchitecture" = x86_gcc2 ]; then
 SECONDARY_ARCHITECTURES="x86"
fi
```
Otherwise, if the ARM arch gets support for using Qemu to run x86 code from packages built for x86 secondary arch then all we would need to do would be to change the test to:
```bash
if [ "$targetArchitecture" = x86_gcc2 ] || [ "$targetArchitecture" = arm ]; then
SECONDARY_ARCHITECTURES="x86"
fi
```
Of course, we could use this instead:
```bash
case "$targetArchitecture" in
x86_gcc2|arm) SECONDARY_ARCHITECTURES="x86";;
esac
```

My conclusion is that x86_64 hybrid **_will_** happen, and it will be much sooner than x86 secondary arch for ARM, if this ever happens.

Reply:
Oops, if we ever need to add x86 secondary arch for ARM+Qemu then this would probably be better:
```diff
-ARCHITECTURES="!x86_gcc2 x86 x86_64"
+ARCHITECTURES="!x86_gcc2 x86 x86_64 !arm"
-SECONDARY_ARCHITECTURES="x86"
+case "$targetArchitecture" in
+x86_gcc2|arm) SECONDARY_ARCHITECTURES="x86";;
+esac
+
+commandSuffix=$secondaryArchSuffix
+commandBinDir=$binDir
+if [ "$targetArchitecture" = x86_gcc2 ]; then
+	commandSuffix=
+	commandBinDir=$prefix/bin
+fi
 
 PROVIDES="
 	slack++$secondaryArchSuffix = $portVersion
-	cmd:slack++$secondaryArchSuffix = $portVersion
+	cmd:slack++$commandSuffix = $portVersion
```
But all this would be totally unneeded if `slack++` builds on ARM.

Reply:
Still, you're talking about adding this for every app recipe. Let's say there is 500 of them. 2 lines x 500 = 1000. Or you can write 10 lines of Python and do that in haikuporter. Bonus: if the condition ever changes you have one place to change, not 500.
You can't have bugs in lines you didn't write.

Reply:
First; there are **only 119 recipes** which meet all the conditions for them to have x86 secondary arch hidden on all arches other than x86_gcc2.
Second, I don't intend to change them in one commit. I will change them when there is a valid reason for changing a recipe.
I'll only make a batch edit when official images of x86_64 hybrid become available, and I'll do that only with those recipes that will not have been "fixed" in the meanwhile.

Moreover, handling these kind of special cases is not a job for haikuporter.

Reply:
>  You can't have bugs in lines you didn't write.

I'm sorry, I don't understand what you are suggesting.

Reply:
He is suggesting that HaikuPorter *is* the correct place to do this ... These are all information that can be gathered from the provides and arch definitions yes??? Then there is no reason *at all* to statically define it....even if doing it in HaikuPorter only covered part of them it would still be the right thing to do....in any case can we just move on from this and move this conversation to a ticket instead of my PR?

Reply:
I have zero interest in impleting this hack so I'm closing it we can revisit it once something has generally been decided on this.